File: dirsync/src/org.ds/l3/OpList.java
package org.ds.l3;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.TreeMap;

import org.ds.l2.DE;
import org.ds.l3.CC.Contents;

@SuppressWarnings("serial")
public class OpList extends HashMap<OpCode, ArrayList<DE>> {
	public static OpList NULL = new OpList();
	public int modcnt;

	public OpList() {
	}

	public void add(OpCode oc, DE de) {
		ArrayList<DE> del = get(oc);
		if(del == null) {
			del = new ArrayList<>();
			put(oc, del);
		}
		del.add(de);
	}

	public OpList(Contents sContents, Contents dContents) {
		TreeMap<String, DE> scf = sContents.files;
		TreeMap<String, DE> scd = sContents.dirs;
		TreeMap<String, DE> dcf = dContents.files;
		TreeMap<String, DE> dcd = dContents.dirs;

		if (sContents.blocked || dContents.blocked) return;
		int sc = sContents.dex ? 1 : 0;
		sc |= dContents.dex ? 2 : 0;
		switch (sc) {
			case 0: // neither exist
				break;
			case 1: // new
				add(OpCode.dCrt, null);
				for (DE de1 : scf.values()) {
					add(OpCode.fNew, de1);
				}
				for (DE de1 : scd.values()) {
					add(OpCode.dNew, de1);
				}
				break;
			case 2: // only dest exists
				if (dContents.dex) { // dir deletion
					for (DE de2 : dcd.values()) {
						add(OpCode.dMis, de2);
					}
					for (DE de2 : dcf.values()) {
						add(OpCode.fMis, de2);
					}
					add(OpCode.dDel, null);
				}
				break;
			case 3: // both exist
				for (DE de2 : dcd.values()) { // deleted dirs
					DE de1 = scd.get(de2.fname);
					if (de1 == null || scf.containsKey(de2.fname)) {
						add(OpCode.dMis, de2);
					}
				}
				for (DE de2 : dcf.values()) { // delete deleted files
					DE de1 = scf.get(de2.fname);
					if (de1 == null || scd.containsKey(de2.fname)) {
						add(OpCode.fMis, de2);
					}
				}
				for (DE de1 : scf.values()) { // possibly copy modified/new files
					DE de2 = dcf.get(de1.fname);
					if (de2 == null) { // new file
						add(OpCode.fNew, de1);
					} else if (!de2.isUp2DateWith(de1)) { // modified file
						add(OpCode.fMod, de1);
					}
				}
				for (DE de1 : scd.values()) { // new/existing dirs
					DE de2 = dcd.get(de1.fname);
					if (de2 == null) { // new
						add(OpCode.dNew, de1);
					} else { // existing
						add(OpCode.dExi, de1);
					}
				}
				break;
		}
	}

	public boolean isEmpty() {
		return size() == 0;
	}
}

File: dircmp/src/org/dc/fn/oplist.go
package fn

import (
	"org/dc/dt"
)

func OpList(sContents, dContents *dt.Contents) (ol dt.OpList) {
	var scf = sContents.Files
	var scd = sContents.Dirs
	var dcf = dContents.Files
	var dcd = dContents.Dirs

	var add = func(oc dt.OpCode, de1 dt.DE) {
		_, ok := ol[oc]
		if !ok {
			ol[oc] = []dt.DE{de1}
		} else {
			ol[oc] = append(ol[oc], de1)
		}
	}

	ol = dt.OpList{}

	if sContents.Blocked || dContents.Blocked {
		if dContents.Blocked {
		}
		return ol
	}

	var sc int
	if sContents.Dex {
		sc |= 1
	}
	if dContents.Dex {
		sc |= 2
	}

	switch sc {
	case 0: // both don't exist
	case 1: // new
		add(dt.DCrt, dt.DE{})
		for _, de1 := range scf {
			add(dt.FNew, de1)
		}
		for _, de1 := range scd {
			add(dt.DNew, de1)
		}
	case 2: // missing
		for _, de2 := range dcd {
			add(dt.DMis, de2)
		}
		for _, de2 := range dcf {
			add(dt.FMis, de2)
		}
		add(dt.DDel, dt.DE{})
	case 3: // both exist
		for _, de2 := range dcd { // delete missing Dirs
			var _, f1 = scd[de2.Fname]
			var _, f2 = scf[de2.Fname]
			if !f1 || f2 {
				add(dt.DMis, de2)
			}
		}
		for _, de2 := range dcf { // delete missing Files
			var _, f1 = scf[de2.Fname]
			var _, f2 = scd[de2.Fname]
			if !f1 || f2 {
				add(dt.FMis, de2)
			}
		}
		for _, de1 := range scf { // possibly copy modified/new Files
			var de2, f1 = dcf[de1.Fname]
			if !f1 { // new file
				add(dt.FNew, de1)
			} else {
				if !de2.IsUpToDateWith(de1) { // modified
					add(dt.FExi, de1)
				}
			}
		}
		for _, de1 := range scd {
			var _, f2 = dcd[de1.Fname]
			if !f2 {
				add(dt.DNew, de1)
			} else {
				add(dt.DExi, de1)
			}
		}
	}
	return ol
}
